From 4d679e0f8ffcb82e76cdb705cd1c11b9eb66a747 Mon Sep 17 00:00:00 2001
From: Kenneth Giusti <kgiusti@gmail.com>
Date: Tue, 19 Jan 2016 15:04:46 -0500
Subject: [PATCH 5/8] fix compilation warnings on RHEL7


diff --git a/plugins/omamqp1/omamqp1.c b/plugins/omamqp1/omamqp1.c
index f184c8f..68d6717 100644
--- a/plugins/omamqp1/omamqp1.c
+++ b/plugins/omamqp1/omamqp1.c
@@ -823,8 +823,8 @@ static void *amqp1_thread(void *arg)
 #if PN_VERSION_MAJOR == 0 && PN_VERSION_MINOR >= 10
         // proton version <= 0.9 did not support Cyrus SASL
         if (cfg->username && cfg->password) {
-            pn_connection_set_user(ps->conn, cfg->username);
-            pn_connection_set_password(ps->conn, cfg->password);
+	    pn_connection_set_user(ps->conn, (char *)cfg->username);
+	    pn_connection_set_password(ps->conn, (char *)cfg->password);
         }
 #endif
         pn_connection_open(ps->conn);
-- 
2.4.3

